crates/extensions/c8y_mapper_ext/src/firmware_update.rs

```diff
@@ -0,0 +1,588 @@
+use crate::converter::CumulocityConverter;
+use crate::error::ConversionError;
+use crate::error::CumulocityMapperError;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestFirmwareRequest;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestRequestGeneric;
+use c8y_api::smartrest::smartrest_serializer::CumulocitySupportedOperations;
+use c8y_api::smartrest::smartrest_serializer::SmartRestSerializer;
+use c8y_api::smartrest::smartrest_serializer::SmartRestSetOperationToExecuting;
+use c8y_api::smartrest::smartrest_serializer::SmartRestSetOperationToFailed;
+use c8y_api::smartrest::smartrest_serializer::SmartRestSetOperationToSuccessful;
+use tedge_api::entity_store::EntityType;
+use tedge_api::messages::FirmwareMetadata;
+use tedge_api::messages::FirmwareUpdateCmdPayload;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::ChannelFilter::Command;
+use tedge_api::mqtt_topics::ChannelFilter::CommandMetadata;
+use tedge_api::mqtt_topics::EntityFilter::AnyEntity;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::OperationType;
+use tedge_api::CommandStatus;
+use tedge_api::Jsonify;
+use tedge_mqtt_ext::Message;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::QoS;
+use tedge_mqtt_ext::TopicFilter;
+use tedge_utils::file::create_directory_with_defaults;
+use tedge_utils::file::create_file_with_defaults;
+use tracing::warn;
+
+pub fn firmware_update_topic_filter(mqtt_schema: &MqttSchema) -> TopicFilter {
+    [
+        mqtt_schema.topics(AnyEntity, Command(OperationType::FirmwareUpdate)),
+        mqtt_schema.topics(AnyEntity, CommandMetadata(OperationType::FirmwareUpdate)),
+    ]
+    .into_iter()
+    .collect()
+}
+
+impl CumulocityConverter {
+    /// Convert c8y_Firmware SmartREST request to ThinEdge firmware_update command.
+    /// Command ID is generated here, but it should be replaced by c8y's operation ID in the future.
+    pub fn convert_firmware_update_request(
+        &self,
+        smartrest: &str,
+    ) -> Result<Vec<Message>, CumulocityMapperError> {
+        if !self.config.capabilities.firmware_update {
+            warn!(
+                "Received a c8y_Firmware operation, however, firmware_update feature is disabled"
+            );
+            return Ok(vec![]);
+        }
```

Comment:
We must check capabilities in `process_smartrest`, otherwise, user cannot use custom operation for `c8y_Firmware` and other operations which are supported by thin-edge.
 
This is the reason why [firmware_operation.robot](https://github.com/thin-edge/thin-edge.io/blob/main/tests/RobotFramework/tests/cumulocity/firmware/firmware_operation.robot) failed in the very first try of this PR. The test uses custom operation for `c8y_Firmware` (`515`). If we don't have the capability check in `process_smartrest` level, the code would enter to this function, instead of calling the custom operation function.

Reply:
Aah okay. Then these warnings also will have to be moved into that `process_smartrest` function, as they are completely useless here.

Reply:
Removed by bfd52a307fbb6768db7308431544b040b30e2c9d